src/representation_theory/rep_equiv.lean

```diff
@@ -0,0 +1,346 @@
+protected theorem map_zero : e 0 = 0 := map_zero e
+@[simp] protected theorem map_smulₛₗ (c : k) (x : V) : e (c • x) = c • e x := e.map_smul' c x
+@[simp] protected theorem map_smulG (g : G) (x : V) : e (ρ g x) = ρ₂ g (e x) := e.map_smulG' g x
+
+@[simp] lemma map_sum {ι : Type*} {s : finset ι} (u : ι → V) :
+  e (∑ i in s, u i) = ∑ i in s, e (u i) := e.to_rep_hom.map_sum
+
+@[simp] theorem map_eq_zero_iff {x : V} : e x = 0 ↔ x = 0 :=
+e.to_add_equiv.map_eq_zero_iff
+theorem map_ne_zero_iff {x : V} : e x ≠ 0 ↔ x ≠ 0 :=
+e.to_add_equiv.map_ne_zero_iff
+
+@[simp] theorem symm_symm (e : ρ ≃ᵣ ρ₂): e.symm.symm = e :=
+by { cases e, refl }
+
+lemma symm_bijective : function.bijective (symm : (ρ ≃ᵣ ρ₂) → (ρ₂ ≃ᵣ ρ)) :=
+equiv.bijective ⟨(symm : (ρ ≃ᵣ ρ₂) →
+  (ρ₂ ≃ᵣ ρ)), (symm : (ρ₂ ≃ᵣ ρ) → (ρ ≃ᵣ ρ₂)), symm_symm, symm_symm⟩
+
+@[simp] lemma mk_coe' (f h₁ h₂ h₃ h₄ h₅) : (rep_equiv.mk f h₁ h₂ h₃ ⇑e h₄ h₅ :
+  ρ₂ ≃ᵣ ρ) = e.symm :=
+symm_bijective.injective $ ext $ λ x, rfl
+
+@[simp] theorem symm_mk (f h₁ h₂ h₃ h₄ h₅) :
+  (⟨e, h₁, h₂, h₃, f, h₄, h₅⟩ : ρ ≃ᵣ ρ₂).symm =
+  { to_fun := f, inv_fun := e,
+    ..(⟨e, h₁, h₂, h₃, f, h₄, h₅⟩ : ρ ≃ᵣ ρ₂).symm } := rfl
+
+@[simp] lemma coe_symm_mk
+  {to_fun inv_fun map_add map_smul map_smulG left_inv right_inv} :
+  ⇑((⟨to_fun, map_add, map_smul, map_smulG, inv_fun, left_inv, right_inv⟩ :
+  ρ ≃ᵣ ρ₂).symm) = inv_fun := rfl
+
+protected lemma bijective : function.bijective e := e.to_equiv.bijective
+protected lemma injective : function.injective e := e.to_equiv.injective
+protected lemma surjective : function.surjective e := e.to_equiv.surjective
+
+protected lemma image_eq_preimage (s : set V) : e '' s = e.symm ⁻¹' s :=
+e.to_equiv.image_eq_preimage s
+
+protected lemma image_symm_eq_preimage (s : set V₂) : e.symm '' s = e ⁻¹' s :=
+e.to_equiv.symm.image_eq_preimage s
+
+section pointwise
+open_locale pointwise
+
+@[simp] lemma image_smul_set (e : ρ ≃ᵣ ρ₂) (c : k) (s : set V) :
+  e '' (c • s) = c • e '' s :=
+rep_hom.image_smul_set e.to_rep_hom c s
+
+@[simp] lemma preimage_smul_set (e : ρ ≃ᵣ ρ₂) (c : k) (s : set V₂) :
+  e ⁻¹' (c • s) = c • e ⁻¹' s :=
+by rw [← rep_equiv.image_symm_eq_preimage, ← rep_equiv.image_symm_eq_preimage,
+  image_smul_set]
+
+end pointwise
+
+-- restrict_scalars
+
+section automorphisms
+
+instance automorphism_group : group (ρ ≃ᵣ ρ) :=
+{ mul := λ f g, g.trans f,
+  one := rep_equiv.refl ρ,
+  inv := λ f, f.symm,
+  mul_assoc := λ f g h, rfl,
+  mul_one := λ f, ext $ λ x, rfl,
+  one_mul := λ f, ext $ λ x, rfl,
+  mul_left_inv := λ f, ext $ f.left_inv }
+
+/-- Restriction from representation automorphisms of `ρ` to representation endomorphisms of `ρ`,
+promoted to a monoid hom. -/
+@[simps]
+def automorphism_group.to_linear_map_monoid_hom : (ρ ≃ᵣ ρ) →* (ρ →ᵣ ρ) :=
+{ to_fun := coe,
+  map_one' := rfl,
+  map_mul' := λ _ _, rfl }
+
+-- apply_distrib_mul_action
+
+end automorphisms
+
+end add_comm_monoid
+
+end rep_equiv
+
+namespace representation
```

Comment:
You shouldn't open empty code blocks like this. You can add a todo comment explaining what needs to be filled out later.